test/python/transpiler/test_consolidate_blocks.py

```diff
@@ -428,6 +429,91 @@ def test_identity_1q_unitary_is_removed(self):
         pm = PassManager([Collect2qBlocks(), Collect1qRuns(), ConsolidateBlocks()])
         self.assertEqual(QuantumCircuit(5), pm.run(qc))
 
+    def test_descent_into_control_flow(self):
+        """Test consolidation in blocks when control flow op is the same as at top level."""
+        qc = QuantumCircuit(2, 1)
+        u2gate1 = U2Gate(-1.2, np.pi)
+        u2gate2 = U2Gate(-3.4, np.pi)
+        qc.append(u2gate1, [0])
+        qc.append(u2gate2, [1])
+        qc.cx(0, 1)
+        qc.cx(1, 0)
+
+        pass_manager = PassManager()
+        pass_manager.append(Collect2qBlocks())
+        pass_manager.append(ConsolidateBlocks(force_consolidate=True))
+        result_top = pass_manager.run(qc)
+
+        qc_control_flow = QuantumCircuit(2, 1)
+        qc_block = copy.deepcopy(qc)
+
+        qc_block = QuantumCircuit(qc.qubits, qc.clbits)
+        qc_block.append(u2gate1, [0])
+        qc_block.append(u2gate2, [1])
+        qc_block.cx(0, 1)
+        qc_block.cx(1, 0)
+
+        ifop = IfElseOp((qc.clbits[0], False), qc_block, None)
+        qc_control_flow.append(ifop, qc.qubits, qc.clbits)
+
+        pass_manager = PassManager()
+        pass_manager.append(Collect2qBlocks())
+        pass_manager.append(ConsolidateBlocks(force_consolidate=True))
+        result_block = pass_manager.run(qc_control_flow)
+        gate_top = result_top[0].operation
+        gate_block = result_block[0].operation.blocks[0][0].operation
+        self.assertEqual(gate_top, gate_block)
```

Comment:
The problem with this test is not inherently the builder interface, it's because the test compares the exact matrix representation of an inner-scope gate with an outer-scope gate without resolving the bit binding of the inner scope.  That's a _vital_ part of resolving control-flow scopes.

It would be good to have tests that explicitly use the builder interface to ensure that there is no problem with bit binding throughout, because that's how users will build control-flow circuits.

Canonicalizing the control-flow structure after the pass has run (which is the only place the test canonicalisation ought to be done, because the tested functionality should work whether or not the form is canonicalised) would not fix this test failure, because the problem is fundamentally that the test was requiring that `A(0, 1)` had the same matrix form as `A(1, 0)`.  Canonicalisation after the pass has run wouldn't change that the operator had been computed on (1, 0) (and suitably transposed) rather than (0, 1).  Instead, the test could check the qubit binding, and swap the 2q operator if necessary:
```python
test_matrix = ...
if <qubit-binding flips order>:
    test_matrix = swap @ test_matrix @ swap
    # or
    test_matrix = test_matrix.reshape(2, 2, 2, 2).transpose(1, 0, 3, 2).reshape(4, 4)
```

Fwiw, the pass is returning a correct answer in the flipped case from what I saw.  It's just the test that's wrong.

Reply:
Is the builder interface the only interface? I mean, is constructing the object directly considered an internal detail? Is bit binding documented anywhere?
If there is a simple interface that is part of the API, it makes sense to me to use this when testing if possible. Unless it is much more difficult than the builder interface.... Also I recall that the builder interface was supposed to be more provisional than the lower-level design.

Reply:
This isn't really about the builder interface, it's about bit-binding that can also be done with the low-level interface.  It's a really important part of the Terra data model that things that interact with scopes (which includes `Instruction.definition` fields and matrices) need to know how to bind the `CircuitInstruction.qubits` to their counterparts in the inner scope.  This existed before the control-flow instructions as well, it's just that it most frequently comes up with them.

The builder interface for control-flow isn't any more provisional than any part of Terra.

Reply:
What is "bit binding"? I don't find anything like that phrase in terra. I really need some documentation or reference to code.

Reply:
By bit binding I mean the relationship between the ordering of `CircuitInstruction.qubits` and how the `operation` is applied. If you access `operation.definition`, or in any way get a matrix representation or inner circuit from the `operation`, you _must_ ensure that you interpret the qubit ordering according to `CircuitInstruction.qubits`. Your test is not doing that - you're taking a 2q matrix without regard to the ordering of the qubits, and consequently it's sometimes got a flipped basis if the block happened to be defined on `1, 0` instead of `0, 1`.

This isn't specific to control flow.

Reply:
And Qiskit does this _everywhere_ - this isn't something new, it's been a fundamental part of the data since the very first versions.